### PR TITLE
Unifying starting reagent init + misc things

### PR DESCRIPTION
**Pull request by PsyCommando**
## Description of changes
For a long while there were like 4 different way to init /obj/item/chem starting reagents. It was really a pain in the ass to deal with. So I merged them together. 

Also, ended up making some changes to welding tools and welder tanks so they're actually /obj/item/chem and don't reimplement chem stuff.
Made the welder pack come with its own integrated welding tool, so its not completely useless and dangerous to wear it.
Reagents dispensers can now be refilled via alt interactions.

## Changelog
:cl:
tweak: Reagent dispensers can now be refilled via alt interactions.
fix: Fixed the welder pack behaving weirdly in inventory and generally being useless.
tweak: Made the welder pack printable, and made it now have an integrated welding tool. Its now much more useful.
refactor: Refactored starting reagents handling on some items.
refactor: Changed how welding tools operate a bit since it used to be a bit all over the place.
refactor: All obj now may implement their initialize_reagents and populate_reagents procs instead of using lists and a million custom implementations for initializing reagents.
/:cl:

### CODE REVIEW

**Comment by noelle-lavenza** [comment]
FWIW, I did something similar to this elsewhere (Aurora and also some other Bayfork?) and I believe it annoyed someone—I don't remember if it was @Lohikar or someone else from Aurora—because every override for `starting_reagents` creates a proc under the hood to initialise the list, and using lists for that kind of thing.

There's also memory concerns, probably—so make sure the list is cleared once it's no longer needed, or maybe switch it to a json-encoded list like with center of mass.

**Comment by Lohikar** [comment]
This is going to create a ton of `(init)` procs, yes -- every type that sets a var to a non-constant value like a list implicitly generates one. I don't like this, but I don't think it's overtly harmful either.

**Comment by PsyCommando** [comment]
Could always just move it all to a proc instead? It would make things simpler, since a bunch things have reagents types and quantities that have to be calculated during init.

I don't particularly like dealing with the initial reagent list anyways, proc would be much easier. It just seemed like that was the most commonly used method, and I assumed that was less likely to trigger pet peeves.

**Comment by PsyCommando** [comment]
So, after dreading to do this for a few weeks I finally went ahead and spent 14hrs of my life to move all initial reagent stuff into procs instead of the list.
I hope you guys are gonna like it.

**Comment by PsyCommando** [comment]
Should be ok for review now. Removing the lists made the diff way larger though.

File: code/datums/extensions/label.dm [hunk 1]
```diff
@@ -41,8 +42,9 @@
 	if(!index) // Playing it safe, something might not have set the name properly
 		return
 
-	user.visible_message("<span class='notice'>\The [user] removes a label from \the [atom_holder].</span>", \
-						 "<span class='notice'>You remove a label, '[label]', from \the [atom_holder].</span>")
+	if(user)
```
Comment (MistakeNot4892) [inline code/datums/extensions/label.dm:45]: Unrelated changes?

File: code/game/machinery/_machines_base/machine_construction/frame.dm [hunk 2]
```diff
@@ -20,7 +20,7 @@
 			machine.anchored = TRUE
 	if(IS_WELDER(I))
 		var/obj/item/weldingtool/WT = I
-		if(!WT.remove_fuel(0, user))
+		if(!WT.weld(0, user))
```
Comment (MistakeNot4892) [inline code/game/machinery/_machines_base/machine_construction/frame.dm]: Really wish you would split distinct changes out into their own PRs, this PR touches 138 files.

File: mods/content/xenobiology/food.dm [hunk 3]
```diff
@@ -55,10 +55,10 @@
 	bitesize = 5
 	eat_sound = 'sound/items/drink.ogg'
 
-/obj/item/chems/food/slimesoup/Initialize()
-	.=..()
+/obj/item/chems/food/slimesoup/populate_reagents()
 	reagents.add_reagent(/decl/material/liquid/slimejelly, 5)
-	reagents.add_reagent(/decl/material/liquid/water, 10)
+	reagents.add_reagent(/decl/material/liquid/slimejelly, 10)
```
Comment (MistakeNot4892) [inline mods/content/xenobiology/food.dm]: Adding 15 slimejelly.

File: code/game/objects/item.dm [hunk 4]
```diff
@@ -942,7 +942,7 @@ modules/mob/living/carbon/human/life.dm if you die, you will be zoomed out.
 		if(istype(source, /obj/structure/reagent_dispensers/watertank))
 			source.reagents.trans_to_obj(src, free_space)
 		else
-			reagents.add_reagent(/decl/material/liquid/water, free_space)
+			reagents.add_reagent(/decl/material/liquid/water, free_space) //#FIXME: Maybe it would be better not to create water from a type check like that in the base item code? Idk.
```
Comment (MistakeNot4892) [inline code/game/objects/item.dm:954]: The only reason it's at /item is that fire extinguishers and water balloons both needed the logic.

File: code/game/objects/items/weapons/electric_welder.dm [hunk 5]
```diff
@@ -71,11 +71,8 @@
 		return
 	. = ..()
 
-/obj/item/weldingtool/electric/burn_fuel(var/amount)
+/obj/item/weldingtool/electric/use_fuel(var/amount)
 	spend_charge(amount * fuel_cost_multiplier)
-	var/turf/T = get_turf(src)
```
Comment (MistakeNot4892) [inline code/game/objects/items/weapons/electric_welder.dm]: Why are you removing this?

File: code/game/objects/items/weapons/paint.dm [hunk 6]
```diff
@@ -18,14 +18,12 @@ var/global/list/cached_icons = list()
 	atom_flags = ATOM_FLAG_OPEN_CONTAINER
 	var/pigment
 
-/obj/item/chems/glass/paint/Initialize()
-	. = ..()
+/obj/item/chems/glass/paint/populate_reagents()
```
Comment (MistakeNot4892) [inline code/game/objects/items/weapons/paint.dm:21]: Huh we did the exact same change in two different PRs, that's kinda funny.

File: code/game/machinery/Sleeper.dm [hunk 7]
```diff
@@ -167,7 +167,8 @@
 			empties++
 			continue
 		var/list/reagent = list()
-		reagent["name"] =   canister.label || "unlabeled"
+		var/datum/extension/labels/lab = get_extension(canister, /datum/extension/labels)
+		reagent["name"] =   (lab?.labels?[1]) || "unlabeled"
```
Comment (MistakeNot4892) [inline code/game/machinery/Sleeper.dm]: This pattern worries me a bit. The compiler sometimes gets confused by chains of `?` and other syntax - would prefer if you broke it out into a couple of checks if possible.

File: code/modules/reagents/reagent_containers/beaker.dm [hunk 8]
```diff
@@ -13,9 +13,9 @@
 	show_reagent_name = TRUE
 	var/lid_color = COLOR_BEASTY_BROWN
 
-/obj/item/chems/glass/beaker/Initialize()
+/obj/item/chems/glass/beaker/examine(mob/user, distance)
 	. = ..()
-	desc += " It can hold up to [volume] units."
+	to_chat(user, " It can hold up to [volume] units.")
```
Comment (MistakeNot4892) [inline code/modules/reagents/reagent_containers/beaker.dm:18]: Note to self to port the Polaris setup where examine() builds a list of strings to join and display.

File: code/modules/reagents/reagent_containers/condiment.dm [hunk 9]
```diff
@@ -177,146 +188,216 @@
...
+/obj/item/chems/condiment/small/soysauce/populate_reagents()
+	reagents.add_reagent(/decl/material/liquid/nutriment/soysauce, reagents.maximum_volume)
 
 //MRE condiments and drinks.
 
 /obj/item/chems/condiment/small/packet
 	icon_state = "packet_small"
 	w_class = ITEM_SIZE_TINY
-	possible_transfer_amounts = @"[1,5,10]"
+	possible_transfer_amounts = @"[1,2,5,10]"
```
Comment (MistakeNot4892) [inline code/modules/reagents/reagent_containers/condiment.dm:238]: My experience with these packets suggests it should be `1,2,3,200`
(this is a joke)

**Comment by MistakeNot4892** [comment]
image
I have gone crosseyed, will review more tomorrow evening or on the weekend.

File: code/game/machinery/_machines_base/machine_construction/frame.dm (hunk 2 above)
Comment (PsyCommando) [reply code/game/machinery/_machines_base/machine_construction/frame.dm]: Yeah, this PR is especially bad for that tbh. I ended up commiting a lot of unrelated stuff in there and I'm not entirely sure why. I could probably try to split off the welder stuff. Only thing is it's gonna take some work to get it working with the old code and I'll have to be super attentive with the conflicts.

File: mods/content/xenobiology/food.dm (hunk 3 above)
Comment (PsyCommando) [reply mods/content/xenobiology/food.dm]: woops

File: code/game/machinery/Sleeper.dm (hunk 7 above)
Comment (PsyCommando) [reply code/game/machinery/Sleeper.dm]: Alright.

File: code/game/objects/items/weapons/paint.dm (hunk 6 above)
Comment (PsyCommando) [reply code/game/objects/items/weapons/paint.dm:21]: wait wut?

File: code/game/objects/items/weapons/electric_welder.dm (hunk 5 above)
Comment (PsyCommando) [reply code/game/objects/items/weapons/electric_welder.dm]: Its already done in the base class, within the weld() proc. welding_tool.dm line 196 iirc.

File: code/game/objects/item.dm (hunk 4 above)
Comment (PsyCommando) [reply code/game/objects/item.dm:954]: I guess I could always make them /obj/item/chem too..

File: code/game/machinery/Sleeper.dm (hunk 7 above)
Comment (PsyCommando) [reply code/game/machinery/Sleeper.dm]: Done!

File: mods/content/xenobiology/food.dm (hunk 3 above)
Comment (PsyCommando) [reply mods/content/xenobiology/food.dm]: Fixed!

**Comment by PsyCommando** [comment]
Well, this was absolute agony to split those things even a little bit. But there you go

**Comment by PsyCommando** [comment]
actually gonna split the welder stuff too. Thought I did..

**Comment by PsyCommando** [comment]
I moved the welder stuff to the welderpack stuff (#2664). But I'm a bit worried some things might have gone lost in the dependency hell. Guess we'll see.

File: code/game/objects/item.dm (hunk 4 above)
Comment (PsyCommando) [reply code/game/objects/item.dm:954]: I made a separate PR with that: #2663

File: code/game/objects/items/weapons/paint.dm (hunk 6 above)
Comment (MistakeNot4892) [reply code/game/objects/items/weapons/paint.dm:21]: I did the same change to the paint logic flow in another commit that I guess didn't get PR'd, just thought it was a funny coincidence.

File: code/game/machinery/Sleeper.dm [hunk 10]
```diff
@@ -167,7 +167,11 @@
 			empties++
 			continue
 		var/list/reagent = list()
-		reagent["name"] =   canister.label || "unlabeled"
+		var/datum/extension/labels/lab = get_extension(canister, /datum/extension/labels)
+		if(lab && (length(lab.labels) > 0))
```
Comment (MistakeNot4892) [inline code/game/machinery/Sleeper.dm]: `length(lab?.labels)` would work here I believe.

File: code/game/objects/item.dm [hunk 11]
```diff
@@ -999,3 +999,11 @@ modules/mob/living/carbon/human/life.dm if you die, you will be zoomed out.
 
 /obj/item/proc/gives_weather_protection()
 	return FALSE
+
+/obj/item/proc/is_used_on(obj/O, mob/user)
```
Comment (MistakeNot4892) [inline code/game/objects/item.dm]: This proc isn't actually used anywhere that I can see.

File: code/game/objects/item.dm [hunk 12]
```diff
@@ -999,3 +999,11 @@ modules/mob/living/carbon/human/life.dm if you die, you will be zoomed out.
...
+	return
+
+/obj/item/proc/updateSelfDialog()
```
Comment (MistakeNot4892) [inline code/game/objects/item.dm:1009]: Neither is this one.

Comment (PsyCommando) [reply code/game/objects/item.dm:1009]: Well, actually, I'm using it in another PR. Not sure why it isn't used more its pretty handy?

File: code/game/objects/item.dm (hunk 11 above)
Comment (PsyCommando) [reply code/game/objects/item.dm]: Dunno, I just moved it over from another file.

Comment (MistakeNot4892) [reply code/game/objects/item.dm]: Might as well delete it then.

File: code/game/objects/objs.dm [hunk 13]
```diff
@@ -249,6 +231,25 @@
 		default_pixel_z = curoff["z"] || 0
 	reset_offsets(0)
 
+/**
+ * Init starting reagents and/or reagent var. Not called at the /obj level.
+ * populate: If set to true, we expect map load/admin spawned reagents to be set.
+ */
+/obj/proc/initialize_reagents(var/populate = TRUE)
+	SHOULD_CALL_PARENT(TRUE)
+	if(reagents.total_volume > 0)
```
Comment (MistakeNot4892) [inline code/game/objects/objs.dm]: Isn't this going to runtime if something hasn't created reagents yet?

File: code/modules/reagents/reagent_dispenser.dm [hunk 14]
```diff
@@ -190,82 +210,107 @@
...
+				break
+			if(istype(C, cup_type))
+				S.remove_from_storage(C, src)
+				qdel(C)
+				cups++
+		return TRUE
+	
 	. = ..()
-	if(!.)
+	if(!. && ATOM_IS_OPEN_CONTAINER(W))
```
Comment (MistakeNot4892) [inline code/modules/reagents/reagent_dispenser.dm:282]: What's the purpose of gating the animation?

Comment (PsyCommando) [reply code/modules/reagents/reagent_dispenser.dm:282]: Because otherwise the animation would play on any unhandled interactions.

File: code/game/objects/objs.dm (hunk 13 above)
Comment (PsyCommando) [reply code/game/objects/objs.dm]: Yup it would.

File: code/game/objects/item.dm (hunk 11 above)
Comment (PsyCommando) [reply code/game/objects/item.dm]: Alright then.

Comment (PsyCommando) [reply code/game/objects/item.dm]: HMMMMMMMMMMMMMMMMMM.. it was not actually unused...
code\game\objects\items\devices\debugger.dm

File: code/modules/reagents/reagent_dispenser.dm (hunk 14 above)
Comment (MistakeNot4892) [reply code/modules/reagents/reagent_dispenser.dm:282]: Shouldn't unhandled interactions be returning false before hitting this point?

Comment (PsyCommando) [reply code/modules/reagents/reagent_dispenser.dm:282]: We don't wanna return when the parent returns false, because reagent_containers interactions are handled in after_attack, not in attackby itself:
```dm
/obj/item/chems/glass/afterattack(var/obj/target, var/mob/user, var/proximity)
	if(!ATOM_IS_OPEN_CONTAINER(src) || !proximity) //Is the container open & are they next to whatever they're clicking?
		return FALSE //If not, do nothing.
	for(var/type in can_be_placed_into) //Is it something it can be placed into?
		if(istype(target, type))
			return TRUE
	if(standard_dispenser_refill(user, target)) //Are they clicking a water tank/some dispenser?
		return TRUE
	if(standard_pour_into(user, target)) //Pouring into another beaker?
		return TRUE
	if(standard_feed_mob(user, target))
		return TRUE
	if(user.a_intent == I_HURT)
		if(standard_splash_mob(user,target))
			return TRUE
		if(reagents && reagents.total_volume)
			to_chat(user, SPAN_DANGER("You splash the contents of \the [src] onto \the [target]."))
			reagents.splash(target, reagents.total_volume)
			return TRUE
	else if(reagents && reagents.total_volume)
		to_chat(user, SPAN_NOTICE("You splash a small amount of the contents of \the [src] onto \the [target]."))
		reagents.splash(target, min(reagents.total_volume, 5))
		return TRUE
	. = ..()
```

Like the intent of that code was to play the bubbles animation when someone fills a glass of water with the machine. But they assumed that only the glass would be unhandled by the attackby, which isn't always the case.

File: code/game/objects/item.dm (hunk 11 above)
Comment (MistakeNot4892) [reply code/game/objects/item.dm]: It's overridden, it's never actually invoked.

File: code/modules/reagents/reagent_dispenser.dm (hunk 14 above)
Comment (MistakeNot4892) [reply code/modules/reagents/reagent_dispenser.dm:282]: Urgh. Attack/afterattack chaining continues to be hellish. Thanks for clarifying.

Comment (PsyCommando) [reply code/modules/reagents/reagent_dispenser.dm:282]: No problems!

File: code/game/objects/item.dm (hunk 11 above)
Comment (PsyCommando) [reply code/game/objects/item.dm]: I just deleted the debugger and the is_used_on proc altogether.

**Comment by MistakeNot4892** [comment]
I am trying so hard to get through this PR.

File: code/modules/reagents/reagent_containers/hypospray.dm [hunk 15]
```diff
@@ -177,31 +198,58 @@
...
+	reagents.add_reagent(/decl/material/liquid/hallucinogenics, reagents.maximum_volume)
+
+////////////////////////////////////////////////////////////////////////////////
+// Autoinjector - Empty
+////////////////////////////////////////////////////////////////////////////////
 /obj/item/chems/hypospray/autoinjector/empty
 	name = "autoinjector"
 	band_color = COLOR_WHITE
-	starts_with = list()
+	material = /decl/material/solid/plastic
```
Comment (MistakeNot4892) [inline code/modules/reagents/reagent_containers/hypospray.dm]: Why are you overriding matter/material specifically on the empty subtype? This should go on the parent if anywhere.

File: code/modules/reagents/reagent_containers/food/sliceable/pizza.dm [hunk 16]
```diff
@@ -134,6 +134,13 @@
 	var/list/boxes = list()// If the boxes are stacked, they come here
 	var/boxtag = ""
 
+/obj/item/pizzabox/Initialize(ml, material_key)
+	. = ..()
+	create_contents()
```
Comment (MistakeNot4892) [inline code/modules/reagents/reagent_containers/food/sliceable/pizza.dm]: This doesn't strike me as related to the reagent stuff.

File: code/modules/reagents/reagent_containers/food/meat/meat.dm [hunk 17]
```diff
@@ -197,10 +197,10 @@
 	center_of_mass = @"{'x':16,'y':16}"
 	bitesize = 3
 
-/obj/item/chems/food/organ/Initialize()
+/obj/item/chems/food/organ/populate_reagents()
 	. = ..()
 	reagents.add_reagent(/decl/material/liquid/nutriment/protein, rand(3,5))
-	reagents.add_reagent(/decl/material/liquid/bromide, rand(1,3))
+	reagents.add_reagent(/decl/material/liquid/bromide,           rand(1,3)) //lolwat?
```
Comment (MistakeNot4892) [inline code/modules/reagents/reagent_containers/food/meat/meat.dm:203]: This is due to replacing generic toxin, I think. Same reason that poison dagger thing in deity code has bromide.

File: code/modules/reagents/reagent_containers/food/junkfood.dm [hunk 18]
```diff
@@ -493,9 +494,9 @@
 	nutriment_amt = 5
 	bitesize = 2
 
-/obj/item/chems/food/weebonuts/Initialize()
+/obj/item/chems/food/weebonuts/populate_reagents()
 	. = ..()
-	reagents.add_reagent(/decl/material/liquid/capsaicin = 1)
+	reagents.add_reagent(/decl/material/liquid/capsaicin, 5)
```
Comment (MistakeNot4892) [inline code/modules/reagents/reagent_containers/food/junkfood.dm]: Original code supplied 1.

File: code/modules/mob/mob.dm [hunk 19]
```diff
@@ -1155,3 +1155,13 @@
 
 /mob/proc/do_flash_animation()
 	return
+
+/mob/proc/unset_machine()
+	src.machine = null
+
+/mob/proc/set_machine(var/obj/O)
```
Comment (MistakeNot4892) [inline code/modules/mob/mob.dm:1163]: Unclear why these changes are in here.

File: code/modules/reagents/reagent_dispenser.dm [hunk 20]
```diff
@@ -190,82 +210,107 @@
 				log_and_message_admins("[key_name_admin(Proj.firer)] shot a fuel tank in \the [area.proper_name].")
 			else
 				log_and_message_admins("shot a fuel tank outside the world.")
-
-		if(!istype(Proj ,/obj/item/projectile/beam/lastertag) && !istype(Proj ,/obj/item/projectile/beam/practice) )
+		
+		if(Proj.damage_flags & DAM_EXPLODE || Proj.damage_type == BURN || Proj.damage_type == ELECTROCUTE || Proj.damage_type == BRUTE)
```
Comment (MistakeNot4892) [inline code/modules/reagents/reagent_dispenser.dm]: `(Proj.damage_flags & DAM_EXPLODE)`

**Comment by MistakeNot4892** [comment]
image

File: code/modules/reagents/reagent_containers/hypospray.dm (hunk 15 above)
Comment (PsyCommando) [reply code/modules/reagents/reagent_containers/hypospray.dm]: Its already on the parent. Probably just a mistake.

File: code/modules/reagents/reagent_containers/food/meat/meat.dm (hunk 17 above)
Comment (PsyCommando) [reply code/modules/reagents/reagent_containers/food/meat/meat.dm:203]: Oh well. That's an odd name for the universal toxin.

File: code/modules/reagents/reagent_containers/food/junkfood.dm (hunk 18 above)
Comment (PsyCommando) [reply code/modules/reagents/reagent_containers/food/junkfood.dm]: Good catch!

File: code/modules/reagents/reagent_containers/food/sliceable/pizza.dm (hunk 16 above)
Comment (PsyCommando) [reply code/modules/reagents/reagent_containers/food/sliceable/pizza.dm]: Probably slipped in while I was doing something else -_-

File: code/modules/mob/mob.dm (hunk 19 above)
Comment (PsyCommando) [reply code/modules/mob/mob.dm:1163]: Basically, those were in a completely random file, so I moved them into the mob.dm file before I forgot about it. 
Its just copy-pasted code as-is.

File: code/game/objects/items/weapons/mop.dm [hunk 21]
```diff
@@ -105,7 +111,7 @@
 	playsound(user, 'sound/machines/click.ogg', 30, 1)
 
 /obj/item/mop/advanced/Process()
-	if(reagents.total_volume < 30)
+	if(reagents.total_volume < MOP_MAX_VOLUME)
```
Comment (MistakeNot4892) [inline code/game/objects/items/weapons/mop.dm]: Why not just `< reagents.max_volume` and remove the need for the define?

File: code/modules/hydroponics/grown_inedible.dm [hunk 22]
```diff
@@ -10,24 +10,30 @@
...
-
-		potency = S.get_trait(TRAIT_POTENCY)
-
-		for(var/rid in S.chems)
-			var/list/reagent_data = S.chems[rid]
-			var/rtotal = reagent_data[1]
-			if(reagent_data.len > 1 && potency > 0)
-				rtotal += round(potency/reagent_data[2])
-			reagents.add_reagent(rid,max(1,rtotal))
+	var/datum/seed/S = SSplants.seeds[plantname]
```
Comment (MistakeNot4892) [inline code/modules/hydroponics/grown_inedible.dm:25]: Why is all this being done in `populate_reagents()` now?

Comment (MistakeNot4892) [reply code/modules/hydroponics/grown_inedible.dm:25]: Think I see. I thought those items kept a reference to their originator but I guess not.

File: code/modules/reagents/reagent_containers/food/meat/meat.dm (hunk 17 above)
Comment (MistakeNot4892) [reply code/modules/reagents/reagent_containers/food/meat/meat.dm:203]: The universal toxin was kind of nonsensical from a material standpoint. Swapping bromide out for more specific toxins would be a good change down the track.

File: code/modules/hydroponics/grown_inedible.dm (hunk 22 above)
Comment (PsyCommando) [reply code/modules/hydroponics/grown_inedible.dm:25]: Well kind of. It keeps the id of the seed that the thing was grown from?

File: code/game/objects/items/weapons/mop.dm (hunk 21 above)
Comment (PsyCommando) [reply code/game/objects/items/weapons/mop.dm]: Truuuuue

**Review by MistakeNot4892 — APPROVED** [review]

**Comment by out-of-phaze** [comment]
I hate this. You now need to use either an annoying proc override or a helper macro to create new chem dispenser cartridges, *and for some stupid fucking reason you undefined it at the end of the file.* You do realize modpacks add those too, right? This entire PR was dumb.
